central/declarativeconfig/manager_impl.go

```diff
@@ -151,14 +190,97 @@ func (m *managerImpl) runReconciliation() {
 }
 
 func (m *managerImpl) reconcileTransformedMessages(transformedMessagesByHandler map[string]protoMessagesByType) {
-	for handler, protoMessagesByType := range transformedMessagesByHandler {
+	log.Debugf("Run reconciliation for the next handlers: %v", maputil.Keys(transformedMessagesByHandler))
+	transformedMessages := map[reflect.Type][]proto.Message{}
+	for _, protoMessagesByType := range transformedMessagesByHandler {
 		for protoType, protoMessages := range protoMessagesByType {
-			// TODO(ROX-14693): Add upserting transformed resources.
-			log.Debugf("Upserting transformed messages of type %s from file %s: %+v",
-				protoType.Name(), handler, protoMessages)
+			transformedMessages[protoType] = append(transformedMessages[protoType], protoMessages...)
 		}
+
 	}
+	m.reconcileUpsertAccessScopes(transformedMessages)
+	m.reconcileUpsertPermissionSets(transformedMessages)
+	m.reconcileUpsertRoles(transformedMessages)
+	m.reconcileUpsertAuthProviders(transformedMessages)
+	m.reconcileUpsertGroups(transformedMessages)
```

Comment:
I think it's relatively easy to hardcode or parameterize order of reconciliation - for now I decided to go with the first approach by introducing list of types in order of reconciliation.

I think what's good about this approach is that by avoiding extra errors we shorten the time for all declarative resources to be reconciled(because if we fail to create a role on the first run, it will only show up in the system on the second run)

Reply:
I agree it's not necessarily required for now.

I'm fine with leaving it for now given the limited number of resources, but in the future we might consider not depending on a specific order if the number of resources grows.